syntax = "proto3";

option optimize_for = LITE_RUNTIME;

import "auth.proto";
import "video.proto";

package aspia.proto;

//
// TODO: Протокол передачи файлов
//

//
// Отправляется только от клиента к серверу
//
message PowerControl
{
    enum Action
    {
        SHUTDOWN  = 0; // Выключение компьютера.
        REBOOT    = 1; // Перезагрузка оперционной системы, в которой запущен сервер
        HIBERNATE = 2; // Переход в режим гибернации
        SUSPEND   = 3; // Переход в режим сна
        LOGOFF    = 4; // Завершение сеанса текущего пользователя
    }

    Action action = 1;
}

//
// Отправляется только от клиента к серверу
//
message KeyEvent
{
    // The keyboard (Caps/Num) lock states.
    enum Flags
    {
        UNKNOWN  = 0;
        CAPSLOCK = 1;
        NUMLOCK  = 2;
        EXTENDED = 4;
        PRESSED  = 8;
    }

    uint32 keycode = 1; // Windows virtual key code
    uint32 flags = 2;
}

//
// Отправляется только от клиента к серверу
//
message PointerEvent
{
    enum ButtonMask
    {
        EMPTY         = 0;
        LEFT_BUTTON   = 1;
        MIDDLE_BUTTON = 2;
        RIGHT_BUTTON  = 4;
        WHEEL_UP      = 8;
        WHEEL_DOWN    = 16;
    }

    uint32 mask = 1; // Button mask
    int32 x = 2;    // x position
    int32 y = 3;    // y position
}

enum ClipboardEncoding
{
    CLIPBOARD_ENCODING_UNKNOWN = 0;
    CLIPBOARD_ENCODING_RAW     = 1;
    CLIPBOARD_ENCODING_ZLIB    = 2;
}

//
// Отправляется в обоих направлениях между сервером и клиентом
//
message Clipboard
{
    ClipboardEncoding encoding = 1;

    string mime_type = 2;
    bytes data = 3;
}

//
// Отправляется только от клиента к серверу
//
message ClipboardControl
{
    enum Flags
    {
        DISABLE_AUTOSEND = 0;
        ENABLE_AUTOSEND  = 1;
        REQUESTED        = 2;
    }

    uint32 flags = 1;
}

enum CursorShapeEncoding
{
    CURSOR_SHAPE_ENCODING_UNKNOWN = 0;
    CURSOR_SHAPE_ENCODING_CACHE   = 1;
    CURSOR_SHAPE_ENCODING_ZLIB    = 2;
}

//
// Отправляется только от сервера к клиенту.
//
message CursorShape
{
    CursorShapeEncoding encoding = 1;

    // Width, height (in screen pixels) of the cursor.
    int32 width  = 2;
    int32 height = 3;

    // X,Y coordinates (relative to upper-left corner) of the cursor hotspot.
    int32 hotspot_x = 4;
    int32 hotspot_y = 5;

    // Cursor pixmap data in 32-bit BGRA format compressed with ZLIB.
    bytes data = 6;

    //
    // Индекс курсора в кеше. Если кодировка не равна CURSOR_SHAPE_ENCODING_CACHE
    // и данный параметр не равен нулю, то кеш должен быть сброшен.
    //
    int32 cache_index = 7;
}

message BellEvent
{
    uint32 flags = 1;
}

//
// Сообщения, которые отправляются от хоста к клиенту
//
message HostToClient
{
    VideoPacket video_packet = 1;
    CursorShape cursor       = 2;
    Clipboard clipboard      = 3;
}

message Control
{
    VideoControl video         = 1;
    ClipboardControl clipboard = 2;
    PowerControl power         = 3;
}

//
// Сообщения, которые отправляются от клиента к хосту
//
message ClientToHost
{
    PointerEvent pointer_event = 1;
    KeyEvent key_event         = 2;
    BellEvent bell_event       = 3;
    Clipboard clipboard        = 4;
    Control control            = 5;
}
